routers/api/packages/generic/generic.go

```diff
@@ -22,11 +23,70 @@ var (
 	filenameRegex    = regexp.MustCompile(`\A[-_+=:;.()\[\]{}~!@#$%^& \w]+\z`)
 )
 
+// PackageFileInfo represents information about an existing package file
+// swagger:model
+type PackageFileInfo struct {
+	// Name of package file
+	Name string `json:"name"`
+	// swagger:strfmt date-time
+	// Date when package file was created/uploaded
+	CreatedUnix timeutil.TimeStamp `json:"created"`
+}
+
+// PackageInfo represents information about an existing package file
+// swagger:model
+type PackageInfo struct {
+	/// Version linked to package information
+	Version string `json:"version"`
+	/// Download count for files within version
+	DownloadCount int64 `json:"downloads"`
+	/// Files uploaded for package version
+	Files []PackageFileInfo `json:"files"`
+}
+
 func apiError(ctx *context.Context, status int, obj any) {
 	message := helper.ProcessErrorForUser(ctx, status, obj)
 	ctx.PlainText(status, message)
 }
 
+// ListPackageVersions lists upload versions and their associated files
+func ListPackageVersions(ctx *context.Context) {
+	pvs, err := packages_model.GetVersionsByPackageName(ctx, ctx.Package.Owner.ID, packages_model.TypeGeneric, ctx.PathParam("packagename"))
+	if err != nil {
+		apiError(ctx, http.StatusInternalServerError, err)
+		return
+	}
+	if len(pvs) == 0 {
+		apiError(ctx, http.StatusNotFound, err)
+		return
+	}
+
+	var info []PackageInfo
+	for _, pv := range pvs {
```

Comment:
It seems strange to list all versions with all files together. IIRC most APIs are designed to "list the versions, and then list the files of a version".

----

What if there are hundreds of versions, and each version has hundreds of files?


----

And it's unclear about how to "for example automatically downloading and deploying the latest compiled build of repos within your organization."

It needs a definition for "what is 'latest'". By `id`, or by time, or by some "tag"?